### ISSUE TITLE
Performance issues on high brush sizes

### ISSUE THREAD

**Issue by t-pearce** [Mavichist/Fountain#9]
Related to my other submitted issue, having a brush in the area of 128 x 128 seems to cause a significant amount of lag, making it hard to make a smooth stroke. My first thought was that the example brushes given in the tutorials might have too high a processing overhead to work over such large an area, but an "eraser" brush that simply returns 0 for both functions also causes a significant input delay.

**Comment by Mavichist** [comment]
With respect; I don't really see these as issues for the purposes of this github. This is for crashes, bugs, errors and usability suggestions. What you seem to be complaining about are the limitations of the software, not issues with its function.

You seem to be asking me to "just make it go faster", as though that idea hasn't occurred to me before. There's a certain amount of processing that needs to be done in order to execute a brush stroke, and naturally that increases exponentially with brush size and linearly with the number of effects you're painting with. The dynamic scripting makes this process a little slower, and that's where your 'eraser' brush suffers its bottleneck, but I'm obviously not going to remove scripting functionality. My advice is to paint without effects, and force a manual update with F5 once you're done painting.

I personally don't find the program to be as slow as you describe it being, and this is the first time someone has decided the lag is so bad they need to consult the author. Please keep in mind that this issue is affected by more than just the code - your system specs are going to influence this to a great degree.

**Comment by t-pearce** [comment]
Performance is definitely a usability issue, at least in my opinion. 

I don't doubt that a certain amount of processing is needed, and I understand the magnitude overheads involved. However, the 'eraser' brush was used without effects; simply a canvas with some random non-0 brush strokes on it, and a brush that returned 0 for all values. Additionally, I'm sure that you have thought of just making it faster, but in my experience, there are always performance improvements that can be made. 

My system specs aren't top-end, I will admit, but I'd expect an AMD FX-6300 w/ 12GB memory and a GTX 760 to be able to cope with this application. Perhaps I'm under the wrong impression there. Still, if you feel the performance here isn't worth investigating further, feel free to close the issue.

**Comment by Mavichist** [comment]
Well firstly let me say that I deliberately abstained from saying 'usability issue' and instead said 'usability suggestion'. Suggesting numerical input in conjunction with existing sliders for color manipulation, as an example of the latter.

What you're talking about is a usability issue, but only when you stretch the program to limits that your system can't cope with very well. I won't close the issue and will try to resolve it, but you need to recognize that, at a point, any performance changes that can be made (and I've made plenty, including changing the scripting language) will have negligible impact. Once major performance updates have been put in place, any subsequent optimizations will do basically nothing by comparison. The impact usually ends up being logarithmic. At this point; there are a couple of things I can do to shave off a small amount of computation, but at a RAM overhead, and they won't really do anything noticeable for the end-user.

Additionally; I find "specs aren't top-end, I'll admit, but I'd expect.. to be able to cope with this application" a bit of a loaded statement. It DOES cope with the application in general, what it doesn't cope with are 8192x8192 renders and large, scripted brushes. I'm not sure why you'd misconstrue this particular set of circumstances with the general use-case of the application.

**Comment by t-pearce** [comment]
While I understand that this is an edge case, and I'm sure you've done a lot of work in optimising the program, it's still a situation which the software is supposed to handle. In order to control somewhat for the hardware side of things, I've found a similarly sized image to the one I was testing on (8218 x 2700, compared to the 8192 x 2048 that I was attempting to create in Fountain). After pasting it into MS Paint, I was able to scroll around the image largely fine. It was a unpleasantly laggy when zoomed out to view the whole image, but that's somewhat to be expected when sampling an image of that size. However, when I zoomed in, scrolling was seamless. Modern iterations of MS Paint also come with fairly complex brush types, and they were able to paint very well both zoomed in and out. Now, obviously, being able to create arbitrary brush types and

**Comment by Mavichist** [comment]
I get the feeling you hadn't finished that post by the time I wrote this, but let me just jump in and ask you; what makes you confident that Fountain is 'supposed to handle' images of that size. Please do not confuse your intentions for the application with the vast majority of end-users.

**Comment by t-pearce** [comment]
Yeah, I accidentally hit ctrl-enter half-way through writing and apparently that posts the comment. So that's good to know.

I suppose my answer to that question would be another question: what makes you confident that Fountain should be considered feature-complete without the ability to handle large images? Further, more on-point to the issue in question here, what makes you confident that Fountain, being able to handle images of least 1024 x 1024, has adequate drawing tools, when the largest stable pen has a width less than 1/8th that of the screen? Are we expected to slowly colour in all the land masses we wish to portray with a small brush?

**Comment by Mavichist** [comment]
Now that I have seen the entire post, I feel I need to comment on this particular point you make:

"Also obviously, MS Paint's likely a lot more optimised than Fountain. This isn't a slight against your programming, but a testament to the pedigree and refinement that Paint's no doubt seen over the years. However, I believe, even accounting for that, there're improvements to be made."

Paint is an incredibly different application - it works very, very differently and it's not even written in the same language. It was written in C++, which is a faster language straight out of the box. Given the language differences, the scripting, the gradient rendering, and every other functional difference that Fountain and Paint have; Fountain would still be a lot slower than Paint even if it had all of the same optimizations (and I'm speaking here largely about the brushing, not the rendering lag) because it needs to do something fundamentally very, very different. It's more or less the apples and oranges scenario.

**Comment by Mavichist** [comment]
I definitely don't mean to sound condescending - in my last post in our other topic I even go so far as to deliberately state that I don't mean to come off as dismissive. Furthermore; I wouldn't be writing responses as detailed and extensive as these (and the ones in the other thread at the moment) if I didn't feel the topic was worthy of dismissal. It's just my frank manner and I promise I don't mean anything by it.

Anyway; I don't consider Fountain feature-complete, even in the vaguest of terms. I have designs for many more features, but this one in particular caters to only a fraction of the user-base and would require an enormous and time-consuming overhaul. Something I feel is unnecessary at this early stage given that I've already given users the blunt instruments necessary to tackle the issue in their own time.

As for the expectation that you paint areas with comparatively small brushes; I don't expect this at all. For painting on a large scale, brushes will be slow, but adjusting the brush power accordingly and tackling broad areas can still be done in a reasonably timely fashion. Using low-frequency noise to brush the broadest of landmass features doesn't take that much time - it's the small detail you really need the high-precision, higher-performance brushes for, and they come in after you've painted the bare, underlying landmass. I'm aware that it's slower than paint; but it's always going to be slower than paint because of the brush scripting. I can't change the C# dynamic assembly, and it's the fastest scripting language I can possibly use because it's native code. The only faster option would be to get people to write their scripts in C, which would be overwhelmingly aggravating, even for me as a person with a programming background.

**Comment by t-pearce** [comment]
Yes, I've just seen the last post in our other topic. I entirely understand, don't worry about it. Just wanted to make sure there wasn't any miscommunication or hostility on either side; us developers can be very protective of our babies!

I also understand not wanting to implement this kind of thing if there are higher priority developments. I feel like I've made my case, either way, in this matter. Whether or not you decide to handle it now or later is entirely up to you.

I have attached a file that illustrates what I mean. It depicts a screenshot of a 128x128 brush, with no effects and no gradient, moved in a smooth circle on a 1024x1024 canvas over the course of approximately one second. The brush in question is the first example brush you give in the readme. You can clearly see where the program's struggled to both render and follow the mouse cursor's movements. You have some of my system specs already, I'm happy to give more if it'd help you. 

I had a bit of a fiddle with the precision value, and a higher precision seemed to help, but you don't actually explain what that does in the readme. Might be a good thing to include.

Would it be possible to limit the render frequency to a lower frame rate than the input handling? That may help the problem.

![image](https://cloud.githubusercontent.com/assets/16254621/18419292/44a56984-784f-11e6-9924-d9520f08576e.png)

**Comment by Mavichist** [comment]
So basically what's going on there is; the program interpolates between the last recorded mouse position and the current mouse position. It derives a path between the two and incurs a paint operation at points along said line. The performance increases when you increase the precision value because the precision value dictates how far apart (in pixels) the paint operations take place from one-another.

I want to ask you a question though; In what set of circumstances will you want to be brushing something with such vigor? Most testing I've done in Fountain involves slowly teasing land up out of the ocean, carefully adding detail and gently smoothing areas. I can't exactly understand why linear artifacts like the one present in that image are relevant to the use-case of the application because I can't imagine a scenario where painting that aggressively is necessary?

**Comment by t-pearce** [comment]
Mostly when blocking out large areas. Say I wanted to create a rough outline for a landmass, I could conceivably get it within alterable tolerances within a few seconds, which could easily create the linear artifacting. 

I suppose my question to you is, where would you want the precision value to be higher than 1? I'm sure there're reasons, because I doubt you'd put it in if there weren't. Dotted lines, perhaps?

**Comment by Mavichist** [comment]
It may seem that way for rendering in black and white, with a brush power that creates hard edges, because the rasterization shows up quite clearly. I can attest, however, that at high resolutions and with larger brushes, it's not at all evident. The brushing was laggy, but I just did this in about 8 seconds. This is a 2048x1024 render and the brush was 256x256.

http://i.imgur.com/nZAkiMY.jpg

Precision values are simply there to help with brush speed. There's no reason to have a precision value of 1 if the brush size is large. Even at a brush size of 64x64; there's little difference between a precision of 1 and a precision of 8 as far as aesthetic goes. The difference as far as performance, however, is huge.
